File: 2021/c.hs
#!/usr/bin/env runghc

-- ACM-ICPC 2021 Domestic Qualifier
-- Problem C: Tree Transformation Puzzle

import Control.Monad (when)
import Control.Monad.State (State, state, runState)

main :: IO ()
main = do
    line <- getLine
    when (line /= "-1") $ do
        let tree = parse line
        print . solve $ convert tree
        main

solve :: MinMaxTree -> Int
solve tree@(Node opv children) =
    max rootv . maximum . map solve' $ rotations children
    where rootv = getMax tree
          solve' (c : cs) = chroot (minMax (opval opv) cs) c

-- compute the max value when a specified non-root node is made the root
-- (pmin, pmax): min and max value of the parent when it becomes a child
chroot :: (Int, Int) -> MinMaxTree -> Int
chroot _ (Node _ []) = minBound
chroot (pmin, pmax) (Node opv children) =
    max rootv . maximum . map chroot' $ rotations children
    where pnode = Node (MinMax {opval = Val 0, minv = pmin, maxv = pmax}) []
          rootv = snd $ minMax (opval opv) (pnode : children)
          chroot' (c : cs) = chroot (minMax (opval opv) (pnode : cs)) c

-- utilities
-- rotations [1,2,3] => [[1,2,3], [2,3,1], [3,1,2]]
rotations :: [a] -> [[a]]
rotations xs = take l . map (take l) . iterate tail $ cycle xs
    where l = length xs

-- trees with the min and max values

data Tree a = Node a [Tree a]  deriving (Show)
data MinMax = MinMax { opval :: ParseLabel, minv :: Int, maxv :: Int }
              deriving (Show)
type MinMaxTree = Tree MinMax

getMin :: MinMaxTree -> Int
getMin (Node mm _) = minv mm
getMax :: MinMaxTree -> Int
getMax (Node mm _) = maxv mm

convert :: ParseTree -> MinMaxTree
convert (Node opval children) =
    let (mn, mx) = minMax opval cs
    in  Node (MinMax {opval = opval, minv = mn, maxv = mx}) cs
    where cs = map convert children

minMax :: ParseLabel -> [MinMaxTree] -> (Int, Int)
minMax opval cs =
    case opval of
        Op Plus  -> (sumMin cs, sumMax cs)
        Op Minus -> (minimum $ map minMinus rotcs,
                     maximum $ map maxMinus rotcs)
        Val v    -> (v, v)
    where rotcs = rotations cs
          sumMin = sum . map getMin
          sumMax = sum . map getMax
          minMinus (hd : tl) = getMin hd - sumMax tl
          maxMinus (hd : tl) = getMax hd - sumMin tl

-- Parser

data Op = Plus | Minus  deriving (Eq, Show)
data ParseLabel = Op Op | Val Int  deriving (Show)
type ParseTree = Tree ParseLabel
type Parser = State String

parse :: String -> ParseTree
parse str = fst $ runState parseRoot str

parseRoot :: Parser ParseTree
parseRoot = do
    l  <- parseNonRoot
    op <- parseOp
    m  <- parseNonRoot
    _  <- parseOp      -- the same as op
    r  <- parseNonRoot
    return $ Node (Op op) [l, m, r]

parseNonRoot :: Parser ParseTree
parseNonRoot = do
    paren <- isNextChar '('
    if paren then do
        getOneChar  -- '('
        l  <- parseNonRoot
        op <- parseOp
        r  <- parseNonRoot
        getOneChar  -- ')'
        return $ Node (Op op) [l, r]
    else do
        v  <- parseVal
        return $ Node (Val v) []

parseOp :: Parser Op
parseOp = do
    ch <- getOneChar
    return $ if ch == '+' then Plus else Minus
parseVal :: Parser Int
parseVal = do
    ch <- getOneChar
    return $ read [ch]

getOneChar :: Parser Char
getOneChar = state $ \s -> (head s, tail s)
isNextChar :: Char -> Parser Bool
isNextChar ch = state $ \s -> (head s == ch, s)

File: 2021/c.exs
#!/usr/bin/env elixir

# ACM-ICPC 2021 Domestic Qualifier
# Problem C - Tree Transformation Puzzle

defmodule Main do
  def main do
    exp = IO.gets("") |> String.trim
    if exp != "-1" do
      solve(exp) |> IO.inspect
      main()
    end
  end

  def solve(exp) do
    ParseTree.parse(exp)    # 構文解析
      |> MinMaxTree.new     # 各部分木の最小値・最大値を計算
      |> MinMaxTree.chroot  # 各頂点を根にしたときの最大値の最大値
  end
end

defmodule Util do
  @doc """
  ```
  iex> rotations [1,2,3]
  [[1,2,3], [2,3,1], [3,1,2]]
  ```
  """
  def rotations(ls) do
    l = length(ls)
    ls |> Stream.cycle |> Stream.iterate(&(Stream.drop(&1, 1)))
       |> Stream.map(&(Enum.take(&1, l))) |> Enum.take(l)
  end
end

# ----------------------------------------
# 構文解析木
defmodule ParseTree do
  defstruct opr: nil, child: []

  # root式
  def parse(exp) do
    {first,  rest} = parseNonRoot(exp)
    {opr,    rest} = parseOpr(rest)
    {second, rest} = parseNonRoot(rest)
    {^opr,   rest} = parseOpr(rest)
    {third,  <<>>} = parseNonRoot(rest)
    %ParseTree{ opr: opr, child: [first, second, third] }
  end

  # non-root式 (先頭が開き括弧)
  defp parseNonRoot(<< ?( :: utf8, rest :: binary >>) do
    {first,  rest} = parseNonRoot(rest)
    {opr,    rest} = parseOpr(rest)
    {second, rest} = parseNonRoot(rest)
    << ?) :: utf8, rest :: binary >> = rest  # 閉じ括弧
    { %ParseTree{ opr: opr, child: [first, second] }, rest }
  end

  # 数字
  defp parseNonRoot(<< digit :: utf8, rest :: binary >>)
         when digit in ?0..?9 do
    { digit - ?0, rest }
  end

  # 演算子
  defp parseOpr(<< opr :: utf8, rest :: binary >>) when opr in [?+, ?-] do
    { (if opr == ?+, do: :plus, else: :minus), rest }
  end
end

# ----------------------------------------
# 最小値・最大値の付いた木
defmodule MinMaxTree do
  defstruct opr: nil, child: [], min: nil, max: nil

  @minBound (-(2 ** 31))  # -♾️

  # ParseTreeからMinMaxTreeに変換
  def new(tree = %ParseTree{}) do
    child = Enum.map(tree.child, &new/1)
    {mn, mx} = minMax(tree.opr, child)
    %MinMaxTree{ opr: tree.opr, child: child, min: mn, max: mx }
  end
  def new(digit), do: %MinMaxTree{ min: digit, max: digit }

  # 演算子opr, 子のリストchildからなる木の最小値・最大値
  defp minMax(opr, child) do
    mnsum = child |> Enum.map(&(&1.min)) |> Enum.sum  # 最小値の総和
    mxsum = child |> Enum.map(&(&1.max)) |> Enum.sum  # 最大値の総和
    mnmxs = child |> Enum.map(&({&1.min, &1.max}))
    if opr == :plus do
      { mnsum, mxsum }
    else  # :minus
      # mn = ((子の1個の最小値) - (それ以外の子の最大値の和)) の最小値
      # mx = ((子の1個の最大値) - (それ以外の子の最小値の和)) の最大値
      mn = mnmxs |> Enum.map(fn {mn, mx} -> mn - (mxsum - mx) end) |> Enum.min
      mx = mnmxs |> Enum.map(fn {mn, mx} -> mx - (mnsum - mn) end) |> Enum.max
      { mn, mx }
    end
  end

  # rootまたはその子孫を根にしたときの最大値.
  # {pmin, pmax}: rootの親を子にしたときの最小値・最大値
  defp chroot(_, %MinMaxTree{child: []}), do: @minBound
  defp chroot({pmin, pmax}, root) do
    parent = %MinMaxTree{min: pmin, max: pmax}  # 親を表す仮頂点
    # rootの子の1個を根にしたときの最大値
    Util.rotations(root.child)
      |> Enum.map(fn [c | cs] -> chroot(minMax(root.opr, [parent|cs]), c) end)
      |> Enum.max
         # rootを根に (rootの子にparentを追加) した場合と比較
      |> max(minMax(root.opr, [parent|root.child]) |> elem(1))
  end
  def chroot(root) do  # rootが元々の根 (子が3個)
    Util.rotations(root.child)
      |> Enum.map(fn [c | cs] -> chroot(minMax(root.opr, cs), c) end)
      |> Enum.max |> max(root.max)
  end
end

# ----------------------------------------
Main.main
